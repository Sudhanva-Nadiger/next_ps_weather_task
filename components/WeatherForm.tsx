'use client'

import React, { ChangeEvent, useRef } from 'react'
import { Input } from './ui/input'
import { Button } from './ui/button'
import { Search } from 'lucide-react'
import { usePathname, useSearchParams, useRouter } from 'next/navigation'

type Props = {}

const WeatherForm = (props: Props) => {
  const ref = useRef<HTMLInputElement>(null)
  const searchParams = useSearchParams()
  const { replace } = useRouter();
  const pathname = usePathname();

  const handleSubmit = (
    e: React.FormEvent<HTMLFormElement>
  ) => {
    e.preventDefault();

    const params = new URLSearchParams(searchParams);
    const term = ref.current?.value;

    if (term) {
      params.set('q', term);
    } else {
      params.delete('q');
    }

    replace(`${pathname}?${params.toString()}`)
  }

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    if(e.target.value === '') {
      const params = new URLSearchParams(searchParams);
      params.delete('q');
      replace(`${pathname}?${params.toString()}`)
    }
  }

  return (
    <form onSubmit={handleSubmit} className='w-full h-fit flex items-end gap-2 shadow border rounded-md p-4'>
      <div className='flex flex-col gap-2 w-full'>
        <label htmlFor='city-input' className='text-start'>City Name</label>
        <Input
          className='peer block w-full rounded-md border border-gray-200 py-[9px] pl-4 text-sm outline-2 placeholder:text-gray-500'
          type='search'
          ref={ref}
          defaultValue={searchParams.get('q')?.toString()}
          placeholder='City name...'
          onChange={handleChange}
        />
      </div>

      <Button type='submit'>
        <Search className='w-4 h-4 mr-2' />
        Search
      </Button>
    </form>
  )
}

export default WeatherForm