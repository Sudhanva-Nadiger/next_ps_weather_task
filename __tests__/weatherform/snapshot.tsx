import '@testing-library/jest-dom/jest-globals';
import '@testing-library/jest-dom';

import WeatherForm from '@/components/WeatherForm';

import { render } from "@testing-library/react";

jest.mock('next/navigation', () => {
  return {
    __esModule: true,
    usePathname: () => ({
      pathname: '',
    }),
    useRouter: () => ({
      push: jest.fn(),
      replace: jest.fn(),
      prefetch: jest.fn(),
    }),
    useSearchParams: () => ({
      get: () => { }
    })
  }
})

describe("Weather Form", () => {

  it("Should render weather form currectly", async () => {
    const { container } = render(<WeatherForm />);
    expect(container).toMatchSnapshot();
  });
})