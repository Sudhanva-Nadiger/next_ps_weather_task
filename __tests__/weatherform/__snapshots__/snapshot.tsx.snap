// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Form Should render weather form currectly 1`] = `
<div>
  <form
    class="w-full h-fit flex items-end gap-2 shadow border rounded-md p-4"
  >
    <div
      class="flex flex-col gap-2 w-full"
    >
      <label
        class="text-start"
        for="city-input"
      >
        City Name
      </label>
      <input
        class="h-10 bg-background px-3 ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 peer block w-full rounded-md border border-gray-200 py-[9px] pl-4 text-sm outline-2 placeholder:text-gray-500"
        placeholder="City name..."
        type="search"
        value=""
      />
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
      type="submit"
    >
      <svg
        class="lucide lucide-search w-4 h-4 mr-2"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <path
          d="m21 21-4.3-4.3"
        />
      </svg>
      Search
    </button>
  </form>
</div>
`;
